### ISSUE TITLE
Http Client is not auto-sending application/x-www-form-urlencoded header

### ISSUE THREAD

**Issue by dakota** [cakephp/cakephp#6840]
According to http://book.cakephp.org/3.0/en/core-libraries/httpclient.html#doing-requests, doing the following should automatically set the `content-type` to `application/x-www-form-urlencoded`

```php
$http = new Client();
$response = $http->post('http://example.com/posts/add', [
  'title' => 'testing',
  'body' => 'content in the post'
]);
```

However, https://github.com/cakephp/cakephp/blob/master/src/Network/Http/Client.php#L429-L431 says otherwise, and https://github.com/cakephp/cakephp/blob/master/src/Network/Http/Adapter/Stream.php#L182-L189 goes even further and forces array form data to be `multipart/form-data` regardless of the `content-type` header set.

We either need to correct the docs, or at least document how to get an `application/x-www-form-urlencoded` request. Alternatively, the code needs to be changed.

**Comment by markstory** [comment]
Is there a problem if `mulitpart/form-data` is used? It defaults to that encoding as the various parts don't know if there are files hiding in the payload and defaults to a value that will always be safe.

I think changing the docs here might be the simplest thing unless there is an issue with defaulting to `multipart/form-data`.

**Comment by davehensley** [comment]
It might be safe to use multipart/form-data for POST requests (if the docs get updated), but an underlying problem with the Stream adapter is that it switches to multipart regardless of what method is being used. So if you make a PUT request and feed it an array (as in the documentation example that @dakota cited), it ends up getting submitted as multipart PUT request.

This is a problem, because PHP (on the receiving end) doesn't know how to handle that, and it may not even be legal under the HTTP spec: https://bugs.php.net/bug.php?id=55815

I've got two Cake apps (one is a RESTful API, and the other is a web app that needs to talk to it). If I use Cake's HTTP client to make a PUT request to the API (and pass it an array as in the documentation's example), the API's $this->request->data is empty. For the moment, I'm working around this by wrapping my data array in http_build_query() before sending it to \Cake\Network\Http\Client::put().

**Comment by ADmad** [comment]
> If I use Cake's HTTP client to make a PUT request to the API (and pass it an array as in the documentation's example), the API's $this->request->data is empty

@lorenzo If you remember similar issue came up in IRC too few days ago. Should we add some magic in RequestHandler to populate request->data for PUT requests too to overcome PHP's limitations?

**Comment by lorenzo** [comment]
@ADmad I think this is a completely different think. But actually we already decode form data (not multipart) into request->data

**Comment by ADmad** [comment]
@lorenzo Okay, guess i remembered incorrectly :)

**Comment by markstory** [comment]
I can look into checking the request payload to see if there are any files before defaulting to multipart/form-data. Would that help solve the problems here?

**Comment by davehensley** [comment]
Yes, that would be great. Also, in \Cake\Network\Http\Adapter\Stream::_buildContent() please check to make sure that the switch to multipart/form-data only occurs when $request->method() == Request::METHOD_POST, because the PUT method uses a different mechanism for uploading a file: http://php.net/manual/en/features.file-upload.put-method.php (although the data is not passed as an array in that case).

**Comment by markstory** [comment]
@davehensley That example is a bit oversimplified, as it doesn't account for file + non-file parameters which again needs multi-part/form-data from what I remember.

**Comment by davehensley** [comment]
Everything I'm reading suggests that multipart/form-data is not intended to be used with PUT (i.e. if you want to combine file+non-file parameters you are supposed to use POST instead). cURL/Guzzle/etc. don't support it on the client side, and PHP itself doesn't support it on the receiving side.

http://www.lornajane.net/posts/2008/accessing-incoming-put-data-from-php#comment-809
https://github.com/appdotnet/api-spec/issues/281
https://groups.google.com/forum/#!topic/guzzle/NDt0av_fRlU
http://stackoverflow.com/questions/15106849/curl-php-uploading-files-with-put-request-how-do-i-process-this

My suggestion would be to avoid this behavior entirely, or at the very least, to make sure that multipart/form-data is NOT used when a file-only PUT request is made.

**Comment by markstory** [comment]
From the guzzle thread, guzzle now supports multipart/form-data requests with PUT, so I see no reason not to support it. We'll continue to bound by the limitation of needing to fit all fileuploads in PHP memory, but that isn't new.

**Comment by markstory** [comment]
Fixed with #6854
